packages/middleware-bucket-endpoint/src/bucketHostnameUtils.ts

```diff
@@ -0,0 +1,128 @@
+import { ARN } from "@aws-sdk/util-arn-parser";
+
+const DOMAIN_PATTERN = /^[a-z0-9][a-z0-9\.\-]{1,61}[a-z0-9]$/;
+const IP_ADDRESS_PATTERN = /(\d+\.){3}\d+/;
+const DOTS_PATTERN = /\.\./;
+export const DOT_PATTERN = /\./;
+export const S3_HOSTNAME_PATTERN = /^(.+\.)?s3[.-]([a-z0-9-]+)\./;
+const S3_US_EAST_1_ALTNAME_PATTERN = /^s3(-external-1)?\.amazonaws\.com$/;
+const AWS_PARTITION_SUFFIX = "amazonaws.com";
+
+export interface AccessPointArn extends ARN {
+  accessPointName: string;
+}
+
+export interface BucketHostnameParameters {
+  baseHostname: string;
+  bucketName: string;
+  accelerateEndpoint?: boolean;
+  dualstackEndpoint?: boolean;
+  pathStyleEndpoint?: boolean;
+  tlsCompatible?: boolean;
+}
+
+export interface ArnHostnameParameters extends Omit<BucketHostnameParameters, "bucketName"> {
```

Comment:
nit: Similar to line 15, should name be shortened to `Params` instead of `Parameters`?
```suggestion
export interface ArnHostnameParams extends Omit<BucketHostnameParams, "bucketName"> {
```